### PR TITLE
Density profiles

### LINKED ISSUE #11: Density profile
An analysis algorithm computing the density profile of some particle along a given axis.

The algorithm should support a profile along a basis axis (x, y, or z); or an axis given by the user from 3 coordinate ([2,2,1] for example).

### PR DESCRIPTION
**Pull request by lscalfi**
Fix #11 
New command to compute density profiles along an axis or radially. Sorry the tests are not totally ready yet.

### CODE REVIEW

**Comment by lscalfi** [comment]
For the radial density, the histogram should be divided by r (since your volume dV is `r.dr.dtheta.dz` in cylindrical coordinates). Should we add it or just output the 'raw' histogram?
Also, for the normalization, maybe we could leave the values without dividing by the max?
Thanks!

**Comment by Luthaf** [comment]
Yes, we should normalize by dividing by r for the radial density.

And we might want to keep the values without dividing by the maximum here, as they have a well defined meaning.

File: src/commands/DensityProfile.cpp [hunk 1]
```diff
@@ -34,7 +36,7 @@ For more information about chemfiles selection language, please see
 Options:
   -h --help                     show this help
   -o <file>, --output=<file>    write result to <file>. This default to the
-                                trajectory file name with the `_dp.dat` extension.
+                                trajectory file name with the `.density.dat` extension.
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: Could you move `extension` to the next line? It looks like it goes after the 80 chars limit.

File: src/commands/DensityProfile.cpp [hunk 2]
```diff
@@ -45,9 +47,11 @@ For more information about chemfiles selection language, please see
...
-                                steps from the input; starting at 0, ending at
-                                the last step, and with a stride of 1.
+                                and <stride> optional. The used steps goes from
+                                <start> to <end> (excluded) by steps of
+                                <stride>. The default values are 0 for <start>,
+                                the number of steps for <end> and 1 for 
+                                <stride>.
   -s <sel>, --selection=<sel>   selection to use for the particles. This must be a
                                 selection of size 1.
                                 [default: atoms: all]
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: Could your move the default to the previous line? It looks weird on its own =)

File: src/Axis.hpp [hunk 3]
```diff
@@ -0,0 +1,57 @@
+// cfiles, an analysis frontend for the Chemfiles library
+// Copyright (C) Guillaume Fraux and contributors -- BSD license
+
+#ifndef CFILES_AXIS_HPP
+#define CFILES_AXIS_HPP
+
+#include <chemfiles.hpp>
+#include "Errors.hpp"
+#include <cmath>
+#include <iostream>
```
Comment (Luthaf) [inline src/Axis.hpp]: You should not need this header here

File: src/Axis.hpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    /// print axis
+    Vector3D& get_coordinates() { return vector_; }
+    bool is_null() { return vector_ == vector3d(0,0,0); }
+
+    /// projection on axis (may be negative)
+    double projection(const Vector3D & positions) {
+        auto dot_product = dot(vector_,positions);
+        return dot_product;
```
Comment (Luthaf) [inline src/Axis.hpp]: You can just `return dot(vector_, position);`

File: src/Axis.hpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        } else if (name=="Y" or name=="y") {
+            vector_ = {0,1,0};
+        } else if (name=="Z" or name=="z") {
+            vector_ = {0,0,1};
+        } else {
+            throw CFilesError("Axis non implemented, enter vector coordinates instead");
+        }
+    }
+    /// Constructor with a 3D vector
+    Axis(double a, double b, double c): vector_(vector3d(a,b,c)) {
```
Comment (Luthaf) [inline src/Axis.hpp]: Can you add a check that the user is not passing 0, 0, 0?

File: src/Axis.hpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        /// normalize the axis
+        vector_ = vector_ / norm(vector_);
+    }
+
+    /// print axis
```
Comment (Luthaf) [inline src/Axis.hpp]: The doc does not correspond to the code =)

File: src/Axis.hpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    Vector3D& get_coordinates() { return vector_; }
+    bool is_null() { return vector_ == vector3d(0,0,0); }
```
Comment (Luthaf) [inline src/Axis.hpp]: Shouldn't this be impossible?

File: src/Axis.hpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    /// projection on axis (may be negative)
+    double projection(const Vector3D & positions) {
+        auto dot_product = dot(vector_,positions);
+        return dot_product;
+    }
+
+    /// radial distance to axis
+    double radial(const Vector3D & positions) {
+        auto dot_product = dot(vector_,positions);
+        auto norme2 = norm2(positions);
```
Comment (Luthaf) [inline src/Axis.hpp]: s/norme2/norm2/

File: src/Histogram.hpp [hunk 9]
```diff
@@ -40,11 +39,16 @@ class Histogram: private std::vector<T> {
     /// Get the size of the bins
     double bin_size() const {return dr_;}
 
+    /// Get the min
```
Comment (Luthaf) [inline src/Histogram.hpp]: => /// Get the minimal value for this histogram

File: src/Histogram.hpp [hunk 10]
```diff
@@ -40,11 +39,16 @@ class Histogram: private std::vector<T> {
...
+    double min() const {return min_;}
+
     /// Insert some `data` in the histogram, and guess the position using the
     /// `bin_size` of the Histogram.
     void insert(T new_data) {
```
Comment (Luthaf) [inline src/Histogram.hpp]: Doc is out of date. Can you rename `new_data` to `data` and remove the part about `bin_size`?

File: src/commands/DensityProfile.cpp [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+For more information about chemfiles selection language, please see
+http://chemfiles.org/chemfiles/latest/selections.html
+
+Usage:
+  cfiles density [options] <trajectory>
+  cfiles density (-h | --help)
+
+Examples:
+  cfiles density water.xyz --cell 15:15:25 --guess-bonds --axis=1,1,1
+  cfiles density in.pdb --selection="x>3"
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: Can you add spaces around '>'? This is not accepted yet.

File: src/commands/DensityProfile.cpp [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+
+For more information about chemfiles selection language, please see
+http://chemfiles.org/chemfiles/latest/selections.html
+
+Usage:
+  cfiles density [options] <trajectory>
+  cfiles density (-h | --help)
+
+Examples:
+  cfiles density water.xyz --cell 15:15:25 --guess-bonds --axis=1,1,1
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: The axis syntax changed

File: src/commands/DensityProfile.cpp (hunk 11 above)
Comment (Luthaf) [reply src/commands/DensityProfile.cpp]: Can you also add examples for radial, origin, points, max and min?

File: src/commands/DensityProfile.cpp [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+                                is <start>:<end>[:<stride>] with <start>, <end>
+                                and <stride> optional. The used steps goes from
+                                <start> to <end> (excluded) by steps of
+                                <stride>. The default values are 0 for <start>,
+                                the number of steps for <end> and 1 for 
+                                <stride>.
+  -s <sel>, --selection=<sel>   selection to use for the particles. This must be a
+                                selection of size 1.
+                                [default: atoms: all]
+  --profile=<str>               type of density profile. Currently implemented are
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: What do you think of having a flag to go from linear to radial? `cfiles density --axis X` would be linear, and `cfiles density --axis X --radial` would be radial.

File: src/commands/DensityProfile.cpp [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    if (args.at("--axis")) {
+        auto splitted = split(args.at("--axis").asString(),':');
+        if (splitted.size() == 1) {
+            axis_ = Axis(args.at("--axis").asString());
+        } else if (splitted.size() == 3) {
+            auto a = string2double(splitted[0]);
+            auto b = string2double(splitted[1]);
+            auto c = string2double(splitted[2]);
+            if (axis_.is_null() ) {
+                throw CFilesError("Null axis does not make sense");
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: does not make sense => does not make sense in density profile

File: src/commands/DensityProfile.cpp [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+            }
+            axis_ = Axis(a,b,c);
+        } else {
+            throw CFilesError("Vector should be of size 3");
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: Vector should => axis for density profile should

File: src/commands/DensityProfile.cpp [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+        }
+    } 
+
+    if (args.at("--origin")) {
+        auto splitted = split(args.at("--origin").asString(),':');
+        if (splitted.size() != 3) {
+            throw CFilesError("The origin option should be a vector of size 3");
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: The origin option should => Origin for density profile should

File: src/commands/DensityProfile.cpp [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+        options_.min = - options_.max;
+        if (args.at("--min")) {
+            options_.min = string2double(args.at("--min").asString());
+        }
+        if (options_.type_profile == "radial") {
+            options_.min = 0;
+        }
+        return Averager<double>(options_.npoints, options_.min, options_.max);
+    } else {
+        throw CFilesError("Please enter a maximum value for the distance values");
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: Please enter a maximum value for distances ~values~ in density profile

File: src/commands/DensityProfile.cpp [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+            std::cout << e.what() << std::endl;
+        }
+    }
+}
+
+void DensityProfile::finish(const Histogram<double>& profile) {
+    std::ofstream outfile(options_.outfile, std::ios::out);
+    if(outfile.is_open()) {
+        outfile << "# Density profile in trajectory " << AveCommand::options().trajectory << std::endl;
+        outfile << "# along axis " << axis_.get_coordinates()[0] << ' ' << axis_.get_coordinates()[1] << ' ';
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: You can write this as: 

```cpp
auto axis = axis_.get_coordinates();
outfile << "# along axis " << axis[0] << ' ' << axis[1] << ' ' << axis[2] << std::endl;
```

File: src/commands/DensityProfile.cpp [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+        outfile << "# Selection: " << options_.selection << std::endl;
+
+        double dr = profile.bin_size();
+        for (size_t i=0; i<profile.size(); i++){
+            if (options_.type_profile == "along_axis") {
+                outfile << profile.min() + i * dr << "  " << profile[i] << "\n";
+            } else if (options_.type_profile == "radial") {
+                auto r = profile.min() + (i + 0.5) * dr;
+                if (r == 0) {
+                    r = dr / 1000;
```
Comment (Luthaf) [inline src/commands/DensityProfile.cpp]: Why the magic constant? Please add a comment to explain this =)

**Review by Luthaf — CHANGES_REQUESTED** [review]
I have quite a few comments on command line interface, feel free to discuss them !

**Comment by Luthaf** [comment]
Also, do not forget to add this to the README!

File: src/Axis.hpp (hunk 7 above)
Comment (lscalfi) [reply src/Axis.hpp]: It will be possible now :)

**Review by Luthaf — APPROVED** [review]

**Pull request merged**